Allow access to clues_metadata during jobs perform method

lwoodson opened the pull request:
- Added Runtime mixin to provide access to jobs that need it
- Store clues_metadata in a threadlocal dupe
- Merge back data from dupe into original on perform_finish
  or failed
- Updated readme with usage
- cleaned up some specs

Added this to allow us to do things like the following example from the README changes:

``` ruby
class PdfGenerationJob
  extend Resque::Plugins::Clues::Runtime

  @queue = pdf_generation
  def self.perform(doc_id)
    doc = Document.find(id)
    clues_metadata[:pdf_generation_strategy] = doc.strategy
    doc.strategy.constantize.new(doc).generate!
  end
end
```

We can then break down our graphs by the strategy used during perform time to see if the job is universally slow or only slow for certain strategies, etc...

Here is a manual verification of this at work.  Define a dummy class like this that injects the clues metadata hash with an environment variable value.

![worker class](http://i.imgur.com/0QMsLJh.png)

Then start two workers with different MESSAGE env variables and make note of their pids.

```
MESSAGE="first worker" QUEUE="test" INTERVAL="0.1" bundle exec rake resque:work &
MESSAGE="second worker" QUEUE="test" INTERVAL="0.1" bundle exec rake resque:work &
```

Open a console and enqueue a few dummy jobs like this:

``` ruby
20.times {Resque.enqueue NoOpWorker}
```

Grep the resque-clues logs for the two env var messages and verify they match the worker pids as expected.

Output for worker 1:

![worker one](http://i.imgur.com/xP3aMb7.png)

Output for worker 2:

![worker two](http://i.imgur.com/FkMItoS.png)

Commit:
Allow access to clues_metadata during jobs perform method

* Added Runtime mixin to provide access to jobs that need it
* Store clues_metadata in a threadlocal dupe
* Merge back data from dupe into original on perform_finish
  or failed
* Updated readme with usage
* cleaned up some specs
* ensure travis builds work on ruby v1.9.3 - v2.1.2

joshuaflanagan commented on lib/resque/plugins/clues/runtime.rb:
```diff
@@ -0,0 +1,72 @@
+          def merge!(original)
+            clues_metadata.keys.each do |key|
+              if original[key].nil?
+                original[key.to_s] = clues_metadata[key]
```
I like that you are protecting from overwriting the 'reserved' values - I wonder if it might be simpler to store the custom data in a nested hash? That way, there is no way to overwrite the reserved values, and developers won't be surprised when a value they write, isn't written (because they weren't aware that it clashed with a reserved key).

Something like:

```
original[:custom_runtime_data] = clues_metadata[:custom_runtime_data]
```

Haven't thought it through completely. Basically, I think the job should have full acces to _read_ all of the metadata, but when it _writes_, it should be constrained to a special bag where it wont conflict with anything.

billeisenhauer commented on README.md:
```diff
@@ -106,11 +109,33 @@ The item is a hash using string keys.  Its structure is as follows:
+
+  @queue = pdf_generation
+  def self.perform(doc_id)
+    doc = Document.find(id)
```
You want to be finding by `doc_id` in this example.

billeisenhauer commented on lib/resque/plugins/clues/runtime.rb:
```diff
@@ -0,0 +1,72 @@
+          end
+
+          ##
+          # Specify the current thread''s clues metadata
```
Extra apostrophe.

billeisenhauer commented on lib/resque/plugins/clues/runtime.rb:
```diff
@@ -0,0 +1,72 @@
+
+          private
+          def set_thread_local(name, val)
+            Thread.current[name.to_sym] = val
```
FYI.  You are `to_sym`ing the name here in these methods, but these are all private methods.  This means that you control how these methods are being called, so you shouldn't need to bother.

billeisenhauer commented on lib/resque/plugins/clues/runtime.rb:
```diff
@@ -0,0 +1,72 @@
+        # will be published with the resque clues
+        # events.
+        def clues_metadata
+          Clues::Runtime.clues_metadata || {}
```
You should defer to the `clues_metadata` class method the knowledge that it must return an empty hash if the hash isn't found in your thread.  There is no guarantee that people will use your convenience method, so it looks like you want them to get the empty hash no matter how they grab for the metadata.

billeisenhauer commented on lib/resque/plugins/clues/runtime.rb:
```diff
@@ -0,0 +1,72 @@
+          # metadata and converting keys to strings.
+          def merge!(original)
+            clues_metadata.keys.each do |key|
+              if original[key].nil?
```
A couple of things:
- You check the original for `key` here, but you `to_s` the key later.  That's inconsistent; is that what you want?
- Might be clearer if you used the dreaded unless to do `unless original.has_key?(key)` assuming you are interested in the presence of the key versus the value being `nil`.

lwoodson replied:
Yeah, a little weirdness here because the hash pulled from redis uses string keys, and someone using this feature may inject data with string or symbol keys.  So it kind of needs a hash with indifferent access type of behavior, but I don't want to tie it to a Rails at all since Resque has no dependencies on rails.  At any rate, making your suggested changes.

lwoodson commented:
Made the changes Bill called out.  Want to talk to Josh a bit on his suggestion.  Also made the travis build work with ruby 1.9.3 to 2.1.2.

lwoodson commented:
Josh raised the issue of trying to namespace the clues system data (like event_hash) from perform runtime data (that a job author could inject).  I agree with the issue, but want to preserve a simple data structure for easy querying via splunk or elastic.  After kicking around the idea a bit, here is what I feel is the best solution:
- Move towards system metadata fields using a '$' prefix.  So 'event_hash' would become '$event_hash'.  
- A side effect of this would be that search dashboards that use 'event_hash' would break
- Have a deprecation process by which we would phase out system metadata fields using the unprefixed names
- Release a version (lets call it A) that contains system metadata in both prefixed and unprefixed fields.  Thus something like 'event_hash' and '$event_hash' would both be populated.  Old dashboards still work.
- Release a subsequent version (lets call it B) that contains only prefixed metadata fields.  Old search dashboards will no longer work, they will have to be modified to use the prefixed metadata fields.
- An upgrade process for someone already using resque clues would be:
  - Use version A for a period of time until data that only had the non-prefixed form is no longer useful.
  - Modify the dashboards to use prefixed field names & verify it works.
  - Upgrade to resque-clues to version B

Am going to push another commit up that basically incorporates everything needed for version A.

joshuaflanagan commented:
I like this plan a lot. Two comments:

1) I think the plan should be that version B in your example is resque-clues 1.0. Version A, with both forms of the data, is still a pre-release version.
2) I would suggest a prefix other than `$`, as it may cause conflicts (or annoying escaping) with query tools like [JSONPath](http://goessner.net/articles/JsonPath/). A single or double underscore `_` prefix might be easier.

Commit:
Move towards system data being stored in attrs prefixed with '_'

* System data will be stored in '_' prefixed attrs
* System data will also be stored in non-prefixed attrs to
  preserve backwards compatability
* Future release will remove the data in the non-prefixed attrs
  allowing people to migrate to v1.2, upgrade their dashboards,
  then migrate to v2.0 to have a seamless upgrade path.

lwoodson commented:
Above done and versioned at 1.0.0-alpha.
